README
Example of use:

page = PAGE

page.10 = IMAGE
page.10.file.import.cObject = BARCODE
page.10.file.import.cObject {
    # UPC exists too but implies prefix=0
    type = EAN13
    backColor = #ffffff
    format = png
    thickness = 1
    prefix = 2
    digits = 099352306753
}

page.20 = IMAGE
page.20.file.import.cObject = BARCODE
page.20.file.import.cObject {
    type = QR

    # Use page URL as content
    content.cObject = TEXT
    content.cObject.typolink {
        parameter.data = TSFE:id
        addQueryString = 1
        addQueryString {
            method = GET
            exclude = id
        }
        forceAbsoluteUrl = 1
        returnLast = url
    }

    # correctionLevel: 'L','M','Q','H'
    correctionLevel = L
    # matrixPointSize: 1-10
    matrixPointSize = 4
}

ChangeLog
2011-05-16  Xavier Perseguers  <xavier@causal.ch>

	* Added support for EAN-13 barcodes
	* Added support for QR codes
	* Initial upload


ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "barcodes".
#
# Auto generated 16-05-2011 10:42
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Barcode Generator',
	'description' => 'This extension generates miscellaneous barcodes.',
	'category' => 'services',
	'author' => 'Xavier Perseguers',
	'author_email' => 'xavier@causal.ch',
	'shy' => '',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'module' => '',
	'state' => 'alpha',
	'internal' => '',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearCacheOnLoad' => 0,
	'lockType' => '',
	'author_company' => 'Causal Sàrl',
	'version' => '0.1.0',
	'constraints' => array(
		'depends' => array(
			'php' => '5.2.0-0.0.0',
			'typo3' => '4.5.0-0.0.0',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => '',
	'suggests' => array(
	),
);

?>

Classes/Services/QR.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2011 Xavier Perseguers (xavier@causal.ch)
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*  A copy is found in the textfile GPL.txt and important notices to the license
*  from the author is found in LICENSE.txt distributed with these scripts.
*
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/


require_once(t3lib_extMgm::extPath('barcodes') . 'Classes/Library/phpqrcode/qrlib.php');

/**
 * Barcode generator, type QR.
 *
 * @category    Services
 * @package     TYPO3
 * @subpackage  tx_barcodes
 * @author      Xavier Perseguers <xavier@causal.ch>
 * @company     Causal Sàrl, http://causal.ch
 * @license     http://www.gnu.org/copyleft/gpl.html
 * @version     SVN: $Id$
 */
class Tx_Barcodes_Services_QR {

	protected $tempPath = 'typo3temp/';
	protected $setup;

	/**
	 * @var tslib_cObj
	 */
	protected $cObj;

	/**
	 * Default constructor.
	 *
	 * @param tslib_cObj $cObj
	 */
	public function __construct(tslib_cObj $cObj) {
		$this->cObj = $cObj;
	}

	/**
	 * Configures the barcode generator.
	 *
	 * @param array $conf
	 * @return void
	 */
	public function start($conf) {
		// Apply stdWrap on each param
		foreach ($conf as $key => $value) {
			if (substr($key, -1) === '.') {
				$baseKey = substr($key, 0, strlen($key) - 1);
				$conf[$baseKey] = $this->cObj->stdWrap($conf[$baseKey], $value);
			}
		}

		if (!trim($conf['content'])) {
			die('No content provided');
		}

		$this->setup = $conf;
	}

	/**
	 * Generates the QR-code.
	 *
	 * @param string $filename
	 * @return void
	 */
	protected function make($filename) {
		$errorCorrectionLevel = $this->setup['correctionLevel'];
		$matrixPointSize = min(max((int)$this->setup['matrixPointSize'], 1), 10);

		QRcode::png($this->setup['content'], $filename, $errorCorrectionLevel, $matrixPointSize, 2);
	}

	/**
	 * Calculates the GIFBUILDER output filename/path based on a serialized, hashed value of this->setup
	 *
	 * @param string Filename prefix, eg. "GB_"
	 * @return string The relative filepath (relative to PATH_site)
	 */
	protected function fileName($pre)  {
		$meaningfulPrefix = '';

		if ($GLOBALS['TSFE']->config['config']['meaningfulTempFilePrefix']) {
			/** @var $basicFileFunctions t3lib_basicFileFunctions */
			$basicFileFunctions = t3lib_div::makeInstance('t3lib_basicFileFunctions');

			$meaningfulPrefix = implode('_', $this->setup['data']);
			$meaningfulPrefix = $basicFileFunctions->cleanFileName($meaningfulPrefix);
			$meaningfulPrefix = substr($meaningfulPrefix, 0, intval($GLOBALS['TSFE']->config['config']['meaningfulTempFilePrefix'])) . '_';
		}

		return $this->tempPath .
				$pre .
				$meaningfulPrefix .
				t3lib_div::shortMD5(serialize($this->setup)) .
				'.png';
	}

	/**
	 * Creates subdirectory in typo3temp/ if not already found.
	 *
	 * @param string Name of sub directory
	 * @return boolean Result of t3lib_div::mkdir(), TRUE if it went well.
	 */
	protected function createTempSubDir($dirName) {

			// Checking if the this->tempPath is already prefixed with PATH_site and if not, prefix it with that constant.
		if (t3lib_div::isFirstPartOfStr($this->tempPath, PATH_site)) {
			$tmpPath = $this->tempPath;
		} else {
			$tmpPath = PATH_site . $this->tempPath;
		}

			// Making the temporary filename:
		if (!@is_dir($tmpPath . $dirName)) {
			return t3lib_div::mkdir($tmpPath . $dirName);
		}
	}

	/**
	 * Initiates the image file generation if ->setup is TRUE and if the file did not exist already.
	 * Gets filename from fileName() and if file exists in typo3temp/ dir it will - of course - not be rendered again.
	 * Otherwise rendering means calling ->make(), then ->output(), then ->destroy()
	 *
	 * @return string The filename for the created GIF/PNG file. The filename will be prefixed "GB_"
	 */
	public function gifBuild() {
		if ($this->setup) {
			$gifFileName = $this->fileName('GB/');	// Relative to PATH_site
			if (!file_exists($gifFileName))	{		// File exists

					// Create temporary directory if not done:
				$this->createTempSubDir('GB/');

					// Create file:
				$this->make($gifFileName);
			}
			return $gifFileName;
		}
	}

}


if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/barcodes/Classes/Services/QR.php']) {
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/barcodes/Classes/Services/QR.php']);
}

?>

Classes/Services/UPC.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2011 Xavier Perseguers (xavier@causal.ch)
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*  A copy is found in the textfile GPL.txt and important notices to the license
*  from the author is found in LICENSE.txt distributed with these scripts.
*
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/


/**
 * Barcode generator, type UPC.
 *
 * @category    Services
 * @package     TYPO3
 * @subpackage  tx_barcodes
 * @author      Xavier Perseguers <xavier@causal.ch>
 * @company     Causal Sàrl, http://causal.ch
 * @license     http://www.gnu.org/copyleft/gpl.html
 * @version     SVN: $Id$
 */
class Tx_Barcodes_Services_UPC extends tslib_gifBuilder {

	protected $x;
	protected $colorOn;
	protected $colorOff;

	/**
	 * @var tslib_cObj
	 */
	protected $cObj;

	/**
	 * Default constructor.
	 *
	 * @param tslib_cObj $cObj
	 */
	public function __construct(tslib_cObj $cObj) {
		$this->cObj = $cObj;
	}

	/**
	 * Configures the barcode generator.
	 *
	 * @param array $conf
	 * @return void
	 */
	public function start(array $conf) {
		// Apply stdWrap on each param
		foreach ($conf as $key => $value) {
			if (substr($key, -1) === '.') {
				$baseKey = substr($key, 0, strlen($key) - 1);
				$conf[$baseKey] = $this->cObj->stdWrap($conf[$baseKey], $value);
			}
		}

		$XY = array(104 * $conf['thickness'], 50);

		if (!preg_match('/^[0-9]{12}$/', $conf['digits'])) {
			die('Digits must contain exactly 12 digits');
		}

		$this->setup = $conf;
		$this->setup['XY'] = $XY[0] . ',' . $XY[1];
	}

	/**
	 * The actual rendering of the image file.
	 * Basically sets the dimensions, the background color, the traverses the array of GIFBUILDER objects and finally setting the transparent color if defined.
	 * Creates a GDlib resource in $this->im and works on that
	 * Called by gifBuild()
	 *
	 * @return	void
	 * @access private
	 */
	public function make() {
		// L/R patterns (http://en.wikipedia.org/wiki/Universal_Product_Code)
		$upcPatterns = array(
			0 => array('L' => '0001101', 'R' => '1110010'),
			1 => array('L' => '0011001', 'R' => '1100110'),
			2 => array('L' => '0010011', 'R' => '1101100'),
			3 => array('L' => '0111101', 'R' => '1000010'),
			4 => array('L' => '0100011', 'R' => '1011100'),
			5 => array('L' => '0110001', 'R' => '1001110'),
			6 => array('L' => '0101111', 'R' => '1010000'),
			7 => array('L' => '0111011', 'R' => '1000100'),
			8 => array('L' => '0110111', 'R' => '1001000'),
			9 => array('L' => '0001011', 'R' => '1110100'),
		);

		$boundaryPatterns = array(
			'S' => '101',
			'M' => '01010',
			'E' => '101',
		);

			// Get trivial data
		$XY = t3lib_div::intExplode(',', $this->setup['XY']);

			// Gif-start
		$this->im = ImageCreateTrueColor($XY[0], $XY[1]);
		$this->w = $XY[0];
		$this->h = $XY[1];

		$this->colorOn = ImageColorAllocate($this->im, 0x00, 0x00, 0x00);
		$this->colorOff = imageColorAllocate($this->im, 0xFF, 0xFF, 0xFF);

			// backColor is set
		$BGcols = $this->convertColor($this->setup['backColor']);
		$Bcolor = ImageColorAllocate($this->im, $BGcols[0], $BGcols[1], $BGcols[2]);
		ImageFilledRectangle($this->im, 0, 0, $this->w, $this->h, $Bcolor);

		$this->x = 0;
		$digits = $this->setup['digits'];

		// show prefix
		$this->drawPattern('00000000', $this->setup['prefix'], floor(.8 * $this->h));
		// start pattern
		$this->drawPattern($boundaryPatterns['S'], '', $this->h);
		for ($i = 0; $i < 6; $i++) {
			$char = $digits{$i};
			$this->drawPattern($upcPatterns[$char]['L'], $char, floor(.8 * $this->h));
		}
		$this->drawPattern($boundaryPatterns['M'], '', $this->h);
		for ($i = 6; $i < 12; $i++) {
			$char = $digits{$i};
			$this->drawPattern($upcPatterns[$char]['R'], $char, floor(.8 * $this->h));
		}
		$this->drawPattern($boundaryPatterns['E'], '', $this->h);

		if ($this->setup['transparentBackground'])	{
				// Auto transparent background is set
			$Bcolor = ImageColorClosest($this->im, $BGcols[0], $BGcols[1], $BGcols[2]);
			ImageColorTransparent($this->im, $Bcolor);
		} elseif (is_array($this->setup['transparentColor_array']))	{
				// Multiple transparent colors are set. This is done via the trick that all transparent colors get converted to one color and then this one gets set as transparent as png/gif can just have one transparent color.
			$Tcolor = $this->unifyColors($this->im, $this->setup['transparentColor_array'], intval($this->setup['transparentColor.']['closest']));
			if ($Tcolor>=0)	{
				ImageColorTransparent($this->im, $Tcolor);
			}
		}
	}

	/**
	 * Draws an UPC barcode pattern.
	 *
	 * @param string $pattern Bit pattern (string with 0's and 1's)
	 * @param string $char Optional character to be printed below
	 * @param integer $height
	 * @return void
	 */
	protected function drawPattern($pattern, $char, $height) {
		if ($char !== '') {
			ImageChar($this->im, 2, $this->x + $this->setup['thickness'], $height - 1, $char, $this->colorOn);
		}
		for ($c = 0; $c < strlen($pattern); $c++) {
			for ($i = 0; $i < $this->setup['thickness']; $i++) {
				$color = $pattern{$c} === '1' ? $this->colorOn : $this->colorOff;
				ImageLine($this->im, $this->x, 0, $this->x, $height, $color);
				$this->x++;
			}
		}
	}

}


if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/barcodes/Classes/Services/UPC.php']) {
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/barcodes/Classes/Services/UPC.php']);
}

?>

Classes/TypoScript/Barcode.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2011 Xavier Perseguers (xavier@causal.ch)
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*  A copy is found in the textfile GPL.txt and important notices to the license
*  from the author is found in LICENSE.txt distributed with these scripts.
*
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/


/**
 * Handles TS content object BARCODE.
 *
 * @category    TypoScript
 * @package     TYPO3
 * @subpackage  tx_barcodes
 * @author      Xavier Perseguers <xavier@causal.ch>
 * @company     Causal Sàrl, http://causal.ch
 * @license     http://www.gnu.org/copyleft/gpl.html
 * @version     SVN: $Id$
 */
class user_barcode {

	/**
	 * Rendering the cObject, BARCODE
	 *
	 * @param string $name Name of the cObject ('BARCODE')
	 * @param array $conf Array of TypoScript properties
	 * @param string $TSkey TS key set to this cObject
	 * @param tslib_cObj $cObj
	 * @return string
	 */
	public function cObjGetSingleExt($name, array $conf, $TSkey, tslib_cObj $cObj) {
		switch ($conf['type']) {
			case 'EAN13':
				$output = $this->getEAN13($conf, $cObj);
				break;
			case 'UPC':
				unset($conf['prefix.']);
				$conf['prefix'] = 0;
				$output = $this->getEAN13($conf, $cObj);
				break;
			case 'QR':
				$output = $this->getQR($conf, $cObj);
				break;
			default:
				$output = 'Invalid type "' . $conf['type'] . '"';
				break;
		}

		return $output;
	}

	/**
	 * Generates an EAN-13 barcode.
	 *
	 * @param array $conf
	 * @param tslib_cObj $cObj
	 * @return string
	 */
	protected function getEAN13(array $conf, tslib_cObj $cObj) {
		/** @var $upc Tx_Barcodes_Services_EAN13 */
		$ean13 = t3lib_div::makeInstance('Tx_Barcodes_Services_EAN13', $cObj);

		$ean13->start($conf);
		return $ean13->gifBuild();
	}

	/**
	 * Generates a QR-code.
	 *
	 * @param array $conf
	 * @param tslib_cObj $cObj
	 * @return string
	 */
	protected function getQR(array $conf, tslib_cObj $cObj) {
		/** @var $upc Tx_Barcodes_Services_QR */
		$qr = t3lib_div::makeInstance('Tx_Barcodes_Services_QR', $cObj);

		$qr->start($conf);
		return $qr->gifBuild();
	}

}


if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/barcodes/Classes/TypoScript/Barcode.php']) {
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/barcodes/Classes/TypoScript/Barcode.php']);
}

?>

Classes/Services/EAN13.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2011 Xavier Perseguers (xavier@causal.ch)
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*  A copy is found in the textfile GPL.txt and important notices to the license
*  from the author is found in LICENSE.txt distributed with these scripts.
*
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/


/**
 * Barcode generator, type EAN-13/UPC.
 *
 * @category    Services
 * @package     TYPO3
 * @subpackage  tx_barcodes
 * @author      Xavier Perseguers <xavier@causal.ch>
 * @company     Causal Sàrl, http://causal.ch
 * @license     http://www.gnu.org/copyleft/gpl.html
 * @version     SVN: $Id$
 */
class Tx_Barcodes_Services_EAN13 extends tslib_gifBuilder {

	protected $x;
	protected $colorOn;
	protected $colorOff;

	/**
	 * @var tslib_cObj
	 */
	protected $cObj;

	/**
	 * Default constructor.
	 *
	 * @param tslib_cObj $cObj
	 */
	public function __construct(tslib_cObj $cObj) {
		$this->cObj = $cObj;
	}

	/**
	 * Configures the barcode generator.
	 *
	 * @param array $conf
	 * @return void
	 */
	public function start(array $conf) {
		// Apply stdWrap on each param
		foreach ($conf as $key => $value) {
			if (substr($key, -1) === '.') {
				$baseKey = substr($key, 0, strlen($key) - 1);
				$conf[$baseKey] = $this->cObj->stdWrap($conf[$baseKey], $value);
			}
		}

		$XY = array(104 * $conf['thickness'], 55);

		if (!preg_match('/^[0-9]$/', $conf['prefix'])) {
			die('Prefix must contain exactly 1 digit');
		}
		if (!preg_match('/^[0-9]{12}$/', $conf['digits'])) {
			die('Digits must contain exactly 12 digits');
		}

		$this->setup = $conf;
		$this->setup['XY'] = $XY[0] . ',' . $XY[1];
	}

	/**
	 * The actual rendering of the image file.
	 * Basically sets the dimensions, the background color, the traverses the array of GIFBUILDER objects and finally setting the transparent color if defined.
	 * Creates a GDlib resource in $this->im and works on that
	 * Called by gifBuild()
	 *
	 * @return	void
	 * @access private
	 */
	public function make() {
		// L/G/R patterns (http://en.wikipedia.org/wiki/EAN-13)
		$ean13Patterns = array(
			0 => array('L' => '0001101', 'G' => '0100111', 'R' => '1110010'),
			1 => array('L' => '0011001', 'G' => '0110011', 'R' => '1100110'),
			2 => array('L' => '0010011', 'G' => '0011011', 'R' => '1101100'),
			3 => array('L' => '0111101', 'G' => '0100001', 'R' => '1000010'),
			4 => array('L' => '0100011', 'G' => '0100011', 'R' => '1011100'),
			5 => array('L' => '0110001', 'G' => '0111001', 'R' => '1001110'),
			6 => array('L' => '0101111', 'G' => '0000101', 'R' => '1010000'),
			7 => array('L' => '0111011', 'G' => '0010001', 'R' => '1000100'),
			8 => array('L' => '0110111', 'G' => '0001001', 'R' => '1001000'),
			9 => array('L' => '0001011', 'G' => '0010111', 'R' => '1110100'),
		);

		// Structure by prefix digit
		$structures = array(
			0 => 'LLLLLLRRRRRR',
			1 => 'LLGLGGRRRRRR',
			2 => 'LLGGLGRRRRRR',
			3 => 'LLGGGLRRRRRR',
			4 => 'LGLLGGRRRRRR',
			5 => 'LGGLLGRRRRRR',
			6 => 'LGGGLLRRRRRR',
			7 => 'LGLGLGRRRRRR',
			8 => 'LGLGGLRRRRRR',
			9 => 'LGGLGLRRRRRR',
		);

		$boundaryPatterns = array(
			'S' => '101',
			'M' => '01010',
			'E' => '101',
		);

			// Get trivial data
		$XY = t3lib_div::intExplode(',', $this->setup['XY']);

			// Gif-start
		$this->im = ImageCreateTrueColor($XY[0], $XY[1]);
		$this->w = $XY[0];
		$this->h = $XY[1];

		$this->colorOn = ImageColorAllocate($this->im, 0x00, 0x00, 0x00);
		$this->colorOff = imageColorAllocate($this->im, 0xFF, 0xFF, 0xFF);

			// backColor is set
		$BGcols = $this->convertColor($this->setup['backColor']);
		$Bcolor = ImageColorAllocate($this->im, $BGcols[0], $BGcols[1], $BGcols[2]);
		ImageFilledRectangle($this->im, 0, 0, $this->w, $this->h, $Bcolor);

		$this->x = 0;
		$digits = $this->setup['digits'];

		// show prefix
		$prefix = $this->setup['prefix'];
		$this->drawPattern('00000000', $prefix, floor(.8 * $this->h));
		// start pattern
		$patternStructure = $structures[$prefix];
		$this->drawPattern($boundaryPatterns['S'], '', $this->h);
		for ($i = 0; $i < 6; $i++) {
			$char = $digits{$i};
			$type = $patternStructure{$i};
			$this->drawPattern($ean13Patterns[$char][$type], $char, floor(.8 * $this->h));
		}
		$this->drawPattern($boundaryPatterns['M'], '', $this->h);
		for ($i = 6; $i < 12; $i++) {
			$char = $digits{$i};
			$type = $patternStructure{$i};
			$this->drawPattern($ean13Patterns[$char][$type], $char, floor(.8 * $this->h));
		}
		$this->drawPattern($boundaryPatterns['E'], '', $this->h);

		if ($this->setup['transparentBackground'])	{
				// Auto transparent background is set
			$Bcolor = ImageColorClosest($this->im, $BGcols[0], $BGcols[1], $BGcols[2]);
			ImageColorTransparent($this->im, $Bcolor);
		} elseif (is_array($this->setup['transparentColor_array']))	{
				// Multiple transparent colors are set. This is done via the trick that all transparent colors get converted to one color and then this one gets set as transparent as png/gif can just have one transparent color.
			$Tcolor = $this->unifyColors($this->im, $this->setup['transparentColor_array'], intval($this->setup['transparentColor.']['closest']));
			if ($Tcolor>=0)	{
				ImageColorTransparent($this->im, $Tcolor);
			}
		}
	}

	/**
	 * Draws an EAN-13 barcode pattern.
	 *
	 * @param string $pattern Bit pattern (string with 0's and 1's)
	 * @param string $char Optional character to be printed below
	 * @param integer $height
	 * @return void
	 */
	protected function drawPattern($pattern, $char, $height) {
		if ($char !== '') {
			ImageChar($this->im, 2, $this->x + $this->setup['thickness'], $height - 1, $char, $this->colorOn);
		}
		for ($c = 0; $c < strlen($pattern); $c++) {
			for ($i = 0; $i < $this->setup['thickness']; $i++) {
				$color = $pattern{$c} === '1' ? $this->colorOn : $this->colorOff;
				ImageLine($this->im, $this->x, 0, $this->x, $height, $color);
				$this->x++;
			}
		}
	}

}


if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/barcodes/Classes/Services/EAN13.php']) {
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/barcodes/Classes/Services/EAN13.php']);
}

?>